man/lpd.8
.ds VE LPRng-3.7.8
.TH LPD 8 \*(VE "LPRng"
.hy 0
.de NP
.IP \\n(nP
.nr nP \\n(nP+1
..
.de np
.nr nP 1
..
.de L
.ie !"\\$1"" \{\
.ft CW
\&\\s-2\\$1\\s+2\\fP\\$2
.ft R
.\}
.el \{\
.ft CW
.ps -2
.\}
..
.ig
lpd.8,v 3.45 1998/03/29 18:37:49 papowell Exp
..
.SH NAME
lpd \- line printer daemon
.SH SYNOPSIS
.B lpd
[\-Llogfile] [\-F] [\-V] [\-Ddebugopt] 
.SH DESCRIPTION
.PP
The lpd program is the printer server program of the LPRng software
suite.
This software is an enhanced and modified version of the Berkeley LPD
software. 
.SH OPTIONS
.IP "\fB\-L\fI logfile" 5
specifies an alternate file to be used for logging error and debugging
messages.
The
.IR syslog (8)
facility is used to log critical messages as well.
.IP "\fB\-F" 5
Under normal operation, the LPD server will run in
background mode.
The -F flag forces it to run in forground mode,
where it is more easily debugged.
.IP "\fB\-V" 5
Print program version information.
.IP "\fB\-D\fI debugopt" 5
Debugging is controlled using the 
.B \-D
option. This accepts a comma-separated list of debugging
settings. These settings take one of two forms:
.B facility=value ,
or
.B value 
to set an overall default value.
The available facilities can be determined by
invoking LPD with the \-D= parameter.
.SH OPERATION
.PP
.I Lpd
is the line printer daemon (spool queue handler) and is normally invoked
at boot time from the
.IR rc (8)
file;
it can also be started by a user.
Note that the lpd server needs only run on systems where actual printing
or spooling is taking place.
.IR lpr (1)
and other related programs transfer files using network facilities to the
.I lpd .
.PP
When started,
.I lpd
reads a configuration file to obtain basic operational parameters
and then reads the
.IR printcap (5)
database information to determine the which printers have spool queues
and to start spool queue server processes.
If running as a background server,
it will disconnect from its control terminal and run in the background.
It uses the system calls
.IR listen (2)
and
.IR accept (2)
to receive requests to print files in the queue,
transfer files to the spooling area,
display the queue,
remove jobs from the queue,
or perform a spool queue control function.
In each case it creates one or more server processes to handle
the request and the lpd process will listen for more requests.
.PP
Sending the server a
SIGHUP signal causes the server to reread the various configuration
and inititialization files.
This action is similar to that of the
.I INETD
and other servers.
The same action is taken when sent a
.I reread
command by the
.IR lpc (1)
program.
At an interveral specified by
the
.I poll_time
configuration variable,
.I lpd
will check for spool queues with jobs and no printing activity,
and start printing.
.PP
LPD access control is done using
a rule set and match algorithm similar to a packet filter.
Each request for printing, status, or control operations
is matched against the rule set,
and the first ACCEPT or REJECT value determines
if the operation can be performed.
The following is a typical permissions file:
.RS
.nf
.L
# Set default permissions
DEFAULT ACCEPT
# Reject any connections from outside our subnet
REJECT SERVICE=X NOT IP=130.191.0.0/255.255.0.0
# Only accept Printing (P) and spooling (LPR) from
# the private network, the 10.0.0.0/8  network and fw
REJECT SERVICE=P,R NOT REMOTEHOST=*.private,10.0.0.0/8,fw.astart.com
# Do not accept forwarded jobs for printing
REJECT SERVICE=P FORWARD
# Allow only the administrators control access
ACCEPT SERVICE=C,M REMOTEHOST=spooler.astart.com USER=root,papowell
ACCEPT SERVICE=C,M SERVER REMOTEUSER=root,papowell
# Allow only the user on the same host who spooled job to remove it
ACCEPT SERVICE=M SAMEUSER SAMEHOST
REJECT SERVICE=M,C
.RE
.fi
.sp
.LP
Permission checking is done by using a set of keys (or fields)
with associated values to check for permission.
The SERVICE key has value P for printing (i.e.- unspooling),
R for spooling (i.e.- LPR request),
C and S for printer control and status respectively (i.e.- LPC request),
M for removal (i.e.- LPRM request),
Q for queue information (i.e.- LPRM request),
and so forth.
The
.l X
key indicates the initial connection to the LPD spooler,
and can be used to control connections from remote systems.
The
values of the USER, HOST, and IP keys taken from the control file
which is being received or checked for permissions.
The REMOTEUSER, REMOTEHOST and REMOTEIP keys are those either sent
as part of a command,
or derived from information about the current network connection.
Each line of the permissions file is scanned for key names and
values, and these are matched against the request keys information.
When all matches on a line are made,
then search terminates with the specified action (ACCEPT/REJECT).
If no match is found the default permission value is used.
The
DEFAULT key is used to specify the current default permission to
be used for successful matches or if there is no match after
scanning the entire permissions database.
.LP
The GROUP entry is used to check that the USER name appears in a
group entry in the system user group database.
For example,
GROUP=student*,staff*
would check to see if any of the group name matching
student* or staff* have the specified user name in them.
If a system has the
.I netgroups
capability,
a group name starting with a
\f(CW@\fR
will be treated as a netgroup name,
and current user name from the job file will bechecked to see if
it is in the group.
Similarly,
the
REMOTEGROUP entry will check a remote user name.
The
PORT
entry can be used to ensure that a connection to the server
orignates from a specified range of ports.
For more details,
see the
.BR lpd.perm (5)
man page.
.LP
The permissions database is scanned in order of the fixed file entries
and then by invoking the specified filters for each
of the permissions lists.
It is recommended that the filters be placed at the end of the
permissions lists.
The user name is one of the parameters passed to the filter,
and can be used to determine if a user has permissions to print a file.
.sp
.nf
.ne 20v
.ta \w'Key__________'u +\w'Match_'u +\w'Connect_'u +\w'Job___'u +\w'Job____'u +\w'LPQ__'u +\w'LPRM__'u +\w'LPC'u
Key	Match	Connect	Job	Job	LPQ	LPRM	LPC
\0	\0	\0	Spool	Print
SERVICE	S	'X'	'R'	'P'	'Q'	'M'	'C,S'
USER	S	-	JUSR	JUSR	JUSR	JUSR	JUSR
HOST	S	RH	JH	JH	JH	JH	JH
GROUP	S	-	JUSR	JUSR	JUSR	JUSR	JUSR
IP	IP	RIP	JIP	JIP	RIP	JIP	JIP
PORT	N	PORT	PORT	-	PORT	PORT	PORT
REMOTEUSER	S	-	JUSR	JUSR	JUSR	CUSR	CUSR
REMOTEHOST	S	RH	RH	JH	RH	RH	RH
REMOTEGROUP	S	-	JUSR	JUSR	JUSR	CUSR	CUSR
REMOTEIP	IP	RIP	RIP	JIP	RIP	RIP	RIP
CONTROLLINE	S	-	CL	CL	CL	CL	CL
PRINTER	S	-	PR	PR	PR	PR	PR
FORWARD	V	-	SA	-	-	SA	SA	SA
SAMEHOST	V	-	SA	-	SA	SA	SA

NEWS
---------------------------------------------------------------------
Version LPRng-3.8.A - not yet released
---------------------------------------------------------------------

This is the first release not done by Patrick Powell but
by the new team to maintain LPRng. The project is hosted
on sourceforge, take a look at
http://sourceforge.net/projects/lprng/
and join us on the new mailing-list at
http://lists.sourceforge.net/mailman/listinfo/lprng-devel

The biggest change is the switch to using automake to generate
the Makefile.im and cleaning the autoconf rules a bit. Most
things should work just as before, with the following exceptions:

- make install no longer tries to remove previous installed
  printing systems and to modify your system to run LPRng, you
  will have to do this yourself. Example files with descriptions
  can be found in the conf/ subdirectory of the source.
- binaries are currently not made suid even if requested,
  that might be reimplemented in the future.
- when configuring with --enable-nls less time is invested in
  finding a gettext implementation. You might need to give it
  some more flags when using a gettext not in your libc.

There is a new configure option --disable-remote to change the default
to not listen via tcp and not accepting requests from remote hosts.

Otherwise a lot of cleanup has happened. Many compiler warnings
quieted, prototypes added, format strings synced with format
of arguments, some code and manpages with strange license situation
replaced, local functions declared static, moving some code around
to not have to link all lpd code in all other binaries...

---------------------------------------------------------------------
Changes in versions until LPRng-3.8.28 can
be found in the file CHANGES.
If that file did not come with your LPRng, you should be able to download
it from: http://lprng.cvs.sourceforge.net/*checkout*/lprng/lprng/CHANGES


DISTRIBUTIONS/FreeBSD-4.ports.sysutils.LPRng/work/LPRng-3.8.3/src/common/lpd.c
/***************************************************************************
 * LPRng - An Extended Print Spooler System
 *
 * Copyright 1988-2001, Patrick Powell, San Diego, CA
 *     papowell@lprng.com
 * See LICENSE for conditions of use.
 *
 ***************************************************************************/

 static char *const _id =
"$Id: lpd.c,v 1.34 2001/12/03 22:08:12 papowell Exp $";


#include "lp.h"
#include "child.h"
#include "fileopen.h"
#include "errorcodes.h"
#include "initialize.h"
#include "linksupport.h"
#include "lpd_logger.h"
#include "getqueue.h"
#include "getopt.h"
#include "proctitle.h"
#if 0
#include "patchlevel.h"
#include "gethostinfo.h"
#include "getprinter.h"
#include "permission.h"
#include "krb5_auth.h"
#include "lpd_secure.h"
#include "lpd_jobs.h"
#include "lpd_dispatch.h"
#endif

/* force local definitions */
#undef EXTERN
#undef DEFINE
#undef DEFS

#define EXTERN
#define DEFINE(X) X
#define DEFS

#include "lpd.h"

/**** ENDINCLUDE ****/

/***************************************************************************
 * main()
 * - top level of LPD Lite.  This is a cannonical method of handling
 *   input.  Note that we assume that the LPD daemon will handle all
 *   of the dirty work associated with formatting, printing, etc.
 * 
 * 1. get the debug level from command line arguments
 * 2. set signal handlers for cleanup
 * 3. get the Host computer Name and user Name
 * 4. scan command line arguments
 * 5. check command line arguments for consistency
 *
 ****************************************************************************/


int main(int argc, char *argv[], char *envp[])
{
	int sock = 0;		/* socket for listen */
	int pid;			/* pid */
	fd_set defreadfds, readfds;	/* for select() */
	struct timeval timeval, *timeout;
	int max_socks;		/* maximum number of sockets */
	int n, m;	/* ACME?  Hmmm... well, ok */
	int err, newsock;
 	time_t last_time;	/* time that last Start_all was done */
 	time_t this_time;	/* current time */
	plp_status_t status;
	int max_servers = 0;
	int start_fd = 0;
	int status_pid = 0;
	int request_pipe[2], status_pipe[2];
	int last_fork_pid_value;
	struct line_list args;
	char *s;
	int first_scan = 1;

	Init_line_list( &args );
	Is_server = 1;	/* we are the LPD server */
	Logger_fd = -1;

#ifndef NODEBUG
	Debug = 0;
#endif
	if(DEBUGL3){
		LOGDEBUG("lpd: argc %d", argc );
		for( n = 0; n < argc; ++n ){
			LOGDEBUG(" [%d] '%s'", n, argv[n] );
		}
		LOGDEBUG("lpd: env" );
		for( n = 0; envp[n]; ++n ){
			LOGDEBUG(" [%d] '%s'", n, envp[n] );
		}
	}

	/* set signal handlers */
	(void) plp_signal(SIGHUP,  (plp_sigfunc_t)Reinit);
	(void) plp_signal(SIGINT, cleanup_INT);
	(void) plp_signal(SIGQUIT, cleanup_QUIT);
	(void) plp_signal(SIGTERM, cleanup_TERM);
	(void) signal(SIGUSR1, SIG_IGN);
	(void) signal(SIGUSR2, SIG_IGN);
	(void) signal(SIGCHLD, SIG_DFL);
	(void) signal(SIGPIPE, SIG_IGN);

	/*
	the next bit of insanity is caused by the interaction of signal(2) and execve(2)
	man signal(2):

	 When a process which has installed signal handlers forks, the child pro-
	 cess inherits the signals.  All caught signals may be reset to their de-
	 fault action by a call to the execve(2) function; ignored signals remain
	 ignored.


	man execve(2):
	 
	 Signals set to be ignored in the calling process are set to be ignored in
					   ^^^^^^^
					   signal(SIGCHLD, SIG_IGN)  <- in the acroread code???
	 
	 the new process. Signals which are set to be caught in the calling pro-  
	 cess image are set to default action in the new process image.  Blocked  
	 signals remain blocked regardless of changes to the signal action.  The  
	 signal stack is reset to be undefined (see sigaction(2) for more informa-
	 tion).


	^&*(*&^!!! &*())&*&*!!!  and again, I say, &*()(&*!!!

	This means that if you fork/execve a child,  then you better make sure
	that you set up its signal/mask stuff correctly.

    So if somebody blocks all signals and then starts up LPD,  it will not work
	correctly.

	*/

	{ plp_block_mask oblock; plp_unblock_all_signals( &oblock ); }


	Get_parms(argc, argv);      /* scan input args */

	Initialize(argc, argv, envp, 'D' );
	DEBUG1("Get_parms: UID_root %d, OriginalRUID %d", UID_root, OriginalRUID);

	if( UID_root && (OriginalRUID != ROOTUID) ){
		FATAL(LOG_ERR) "lpd installed SETUID root and started by user %d! Possible hacker attack", OriginalRUID);
	}

	Setup_configuration();

#if 0
	if( Worker_LPD ){
		struct line_list args;
		Name = "LPD_WORKER";
		DEBUG1("Lpd_worker: argc %d, optind %d", argc, Optind );
		Init_line_list( &args );
		while( Optind < argc ){
			Add_line_list(&args,argv[Optind++],Value_sep,1,1);
		}
		if(DEBUGL1)Dump_line_list("Lpd_worker - args", &args );
		Do_work(0,Worker_LPD,args);
		cleanup(0);
	}
#endif

	/* get the maximum number of servers allowed */
	max_servers = Get_max_servers();
	n = Get_max_fd();
	DEBUG1( "lpd: maximum servers %d, maximum file descriptors %d ",
		max_servers, n );

	if( Lockfile_DYN == 0 ){
		LOGERR_DIE(LOG_INFO) _("No LPD lockfile specified!") );
	}

	/* chdir to the root directory */
	if( chdir( "/" ) == -1 ){
		Errorcode = JABORT;
		LOGERR_DIE(LOG_ERR) "cannot chdir to /");
	}

	/*
	 * This is the one of the two places where we need to be
	 * root in order to open a socket
	 */

	if( Lpd_port_arg ){
		Set_DYN( &Lpd_port_DYN, Lpd_port_arg );
	}
	sock = Link_listen();
	DEBUG1("lpd: listening socket fd %d",sock);
	if( sock < 0 ){
		/*
		 * try reading the lockfile
		 */
  		pid = Get_lpd_pid();
 		if( pid > 0 && kill(pid,0) ) pid = 0;
		Errorcode = 1;
  		if( pid > 0 ){
  			DIEMSG( _("Another print spooler is using TCP printer port, possibly lpd process '%d'"),
  				pid );
  		} else {
 			if( !UID_root ){
 				DIEMSG("Not running with ROOT perms and trying to open port '%s'", Lpd_port_DYN );
 			}
 			DIEMSG( _("Another print spooler is using TCP printer port - not LPRng") );
  		}
		Errorcode = JSUCC;
	}

	/* setting nonblocking on the listening fd
	 * will prevent a problem with terminations of connections
	 * before ACCEPT has completed
	 *  1. user connects, does the 3 Way Handshake
	 *  2. before accept() is done,  a RST packet is sent
	 *  3. a select() will succeed, but the accept() will hang
	 *  4. if the non-blocking mode is used, then the select will
	 *     succeed and the accept() will fail
	 */
	Set_nonblock_io(sock);

	/*
	 * At this point you are the server for the LPD port
	 * you need to fork to allow the regular user to continue
	 * you put the child in its separate process group as well
	 */
	Name = "MAIN";
	setproctitle( "lpd %s", Name  );

	if( (pid = dofork(1)) < 0 ){
		LOGERR_DIE(LOG_ERR) _("lpd: main() dofork failed") );
	} else if( pid ){
		if( Foreground_LPD ){
			while( (pid = plp_waitpid( pid, &status, 0)) > 0 ){
				DEBUG1( "lpd: process %d, status '%s'",
					pid, Decode_status(&status));
			}
		}
		Errorcode = 0;
		exit(0);
	}

	/* set up the log file and standard environment - do not
	   fool around with anything but fd 0,1,2 which should be safe
		as we made sure that the fd 0,1,2 existed.
    */

	Setup_log( Logfile_LPD );

	Name = "Waiting";
	setproctitle( "lpd %s", Name  );

	/*
	 * Write the PID into the lockfile
	 */

	Set_lpd_pid();

	if( Drop_root_DYN ){
		Full_daemon_perms();
	}

	/* establish the pipes for low level processes to use */
	if( pipe( request_pipe ) == -1 ){
		LOGERR_DIE(LOG_ERR) _("lpd: pipe call failed") );
	}
	Max_open(request_pipe[0]); Max_open(request_pipe[1]);
	DEBUG2( "lpd: fd request_pipe(%d,%d)",request_pipe[0],request_pipe[1]);
	Lpd_request = request_pipe[1];
	Set_nonblock_io( Lpd_request );

	Logger_fd = -1;
	status_pid = -1;
	if( Logger_destination_DYN ){
		if( pipe( status_pipe ) == -1 ){
			LOGERR_DIE(LOG_ERR) _("lpd: pipe call failed") );
		}
		Max_open(status_pipe[0]); Max_open(status_pipe[1]);
		Logger_fd = status_pipe[1];
		DEBUG2( "lpd: fd status_pipe(%d,%d)",status_pipe[0],status_pipe[1]);
		status_pid = Start_logger( status_pipe[0] );
		if( status_pid < 0 ){
			LOGERR_DIE(LOG_ERR) _("lpd: cannot start initial logger process") );
		}
	}

	/* open a connection to logger */
	setmessage(0,LPD,"Starting");

	/* get the maximum number of servers allowed */
	max_servers = Get_max_servers();
	DEBUG1( "lpd: maximum servers %d", max_servers );

	/*
	 * clean out the current queues
	 */
 	last_time = time( (void *)0 );

	/* set up the wait activity */

	FD_ZERO( &defreadfds );
	FD_SET( sock, &defreadfds );
	FD_SET( request_pipe[0], &defreadfds );

	/*
	 * start waiting for connections from processes
	 */

	last_time = time( (void *)0 );
	last_fork_pid_value = start_fd = Start_all(first_scan);
	Fork_error( last_fork_pid_value );
	if( start_fd > 0 ){
		first_scan = 0;
	}

	do{
		DEBUG1("lpd: LOOP START");
		if(DEBUGL3){ int fd; fd = dup(0); LOGDEBUG("lpd: next fd %d",fd); close(fd); };

		timeout = 0;
		DEBUG2( "lpd: Poll_time %d, Force_poll %d, start_fd %d, Started_server %d",
			Poll_time_DYN, Force_poll_DYN, start_fd, Started_server );
		if(DEBUGL2)Dump_line_list("lpd - Servers_line_list",&Servers_line_list );

		/*
		 * collect zombies 
		 */

		while( (pid = plp_waitpid( -1, &status, WNOHANG)) > 0 ){
			DEBUG1( "lpd: process %d, status '%s'",
				pid, Decode_status(&status));
			if( pid == status_pid ){
				status_pid = -1;
			}
			last_fork_pid_value = 1;
		}
		if( last_fork_pid_value > 0 && Logger_fd > 0 && status_pid < 0 ){
			DEBUG1( "lpd: restarting logger process");
			last_fork_pid_value = status_pid = Start_logger( status_pipe[0] );
			Fork_error( last_fork_pid_value );
			DEBUG1("lpd: status_pid %d", status_pid );
		}
		if( last_fork_pid_value < 0 ){
			/* wait for 10 seconds then go in a loop */
			memset(&timeval, 0, sizeof(timeval));
			timeval.tv_sec = 10;
			timeout = &timeval;
		} else if( Poll_time_DYN > 0
			&& start_fd <= 0 && Servers_line_list.count == 0 ){
			memset(&timeval, 0, sizeof(timeval));
			this_time = time( (void *)0 );
			m = (this_time - last_time);
			timeval.tv_sec = Poll_time_DYN - m;
			timeout = &timeval;
			DEBUG2("lpd: from last poll %d, to next poll %d",
				m, (int)timeval.tv_sec );
			if( m >= Poll_time_DYN ){
				if( Started_server || Force_poll_DYN ){
					last_fork_pid_value = start_fd = Start_all(first_scan);
					Fork_error( last_fork_pid_value );
					DEBUG1( "lpd: restarting poll, start_fd %d", start_fd);
					if( start_fd > 0 ){
						if( first_scan ) first_scan = 0;
						last_time = this_time;
						timeval.tv_sec = Poll_time_DYN;
					}
				} else {
					DEBUG1( "lpd: no poll" );
					timeout = 0;
				}
			}
		}

		n = Countpid();
		max_servers = Get_max_servers();
		DEBUG1("lpd: max_servers %d, active %d", max_servers, n );

		/* allow a little space for people to send commands */
		while( last_fork_pid_value > 0 && Servers_line_list.count > 0 && n < max_servers-4 ){ 
			s = Servers_line_list.list[0];
			DEBUG1("lpd: starting server '%s'", s );
			Set_str_value(&args,PRINTER,s);
			last_fork_pid_value = pid = Start_worker( "queue",&args, 0 );
			Fork_error( last_fork_pid_value );
			Free_line_list(&args);
			if( pid > 0 ){
				Remove_line_list( &Servers_line_list, 0 );
				++Started_server;
				++n;
			}
		}

		DEBUG1("lpd: last_fork_pid_value %d, processes %d active, max %d",
			last_fork_pid_value, n, max_servers );
		/* do not accept incoming call if no worker available */
		readfds = defreadfds;
		if( n >= max_servers || last_fork_pid_value < 0 ){
			DEBUG1( "lpd: not accepting requests" );
			FD_CLR( sock, &readfds );
		}

		max_socks = sock+1;
		if( request_pipe[0] >= max_socks ){
			max_socks = request_pipe[0]+1;
		}
		if( start_fd > 0 ){
			FD_SET( start_fd, &readfds );
			if( start_fd >= max_socks ){
				max_socks = start_fd+1;
			}
		}

		DEBUG1( "lpd: starting select timeout '%s', %d sec, max_socks %d",
		timeout?"yes":"no", (int)(timeout?timeout->tv_sec:0), max_socks );
		if(DEBUGL2){
			int i;
			for(i=0; i < max_socks; ++i ){
				if( FD_ISSET( i, &readfds ) ){
					LOGDEBUG( "lpd: waiting for fd %d to be readable", i );
				}
			}
		}
		Setup_waitpid_break();
		errno = 0;
		m = select( max_socks,
			FD_SET_FIX((fd_set *))&readfds,
			FD_SET_FIX((fd_set *))0,
			FD_SET_FIX((fd_set *))0, timeout );
		err = errno;
		Setup_waitpid();
		if(DEBUGL1){
			int i;
			LOGDEBUG( "lpd: select returned %d, error '%s'",
				m, Errormsg(err) );
			for(i=0; i < max_socks; ++i ){
				if( FD_ISSET( i, &readfds ) ){
					LOGDEBUG( "lpd: fd %d readable", i );
				}
			}
		}
		/* if we got a SIGHUP then we reread configuration */
		if( Reread_config || !Use_info_cache_DYN ){
			DEBUG1( "lpd: rereading configuration" );
			/* we need to force the LPD logger to use new printcap information */
			if( Reread_config ){
				if( status_pid > 0 ) kill( status_pid, SIGINT );
				setmessage(0,LPD,"Restart");
				Reread_config = 0;
			}
			Setup_configuration();
		}
		/* mark this as a timeout */
		if( m < 0 ){
			if( err != EINTR ){
				errno = err;
				LOGERR_DIE(LOG_ERR) _("lpd: select error!"));
				break;
			}
			continue;
		} else if( m == 0 ){
			DEBUG1( "lpd: signal or time out, last_fork_pid_value %d", last_fork_pid_value );
			/* we try to fork now */
			if( last_fork_pid_value < 0 ) last_fork_pid_value = 1;
			continue;
		}
		if( FD_ISSET( sock, &readfds ) ){
			struct sockaddr sinaddr;
			int len;
			len = sizeof( sinaddr );
			newsock = accept( sock, &sinaddr, &len );
			err = errno;
			DEBUG1("lpd: connection fd %d", newsock );
			if( newsock > 0 ){
				pid = Start_worker( "server", &args, newsock );
				if( pid < 0 ){
					LOGERR(LOG_INFO) _("lpd: fork() failed") );
					Write_fd_str( newsock, "\002Server load too high\n");
				} else {
					DEBUG1( "lpd: listener pid %d running", pid );
				}
				close( newsock );
				Free_line_list(&args);
			} else {
				errno = err;
				LOGERR(LOG_INFO) _("Service_connection: accept on listening socket failed") );
			}
		}
		if( FD_ISSET( request_pipe[0], &readfds ) 
			&& Read_server_status( request_pipe[0] ) == 0 ){
			Errorcode = JABORT;
			LOGERR_DIE(LOG_ERR) _("lpd: Lpd_request pipe EOF! cannot happen") );
		}
		if( start_fd > 0 && FD_ISSET( start_fd, &readfds ) ){
			start_fd = Read_server_status( start_fd );
		}
	}while( 1 );
	Free_line_list(&args);
	cleanup(0);
	return(0);
}

/***************************************************************************
 * Setup_log( char *logfile, int sock )
 * Purpose: to set up a standard error logging environment
 * saveme will prevent STDIN from being clobbered
 *   1.  dup 'sock' to fd 0, close sock
 *   2.  opens /dev/null on fd 1
 *   3.  If logfile is "-" or NULL, output file is alread opened
 *   4.  Open logfile; if unable to, then open /dev/null for output
 ***************************************************************************/
void Setup_log(char *logfile )
{
	struct stat statb;

	close(0); close(1);
	if (open("/dev/null", O_RDONLY, 0) != 0) {
	    LOGERR_DIE(LOG_ERR) _("Setup_log: open /dev/null failed"));
	}
	if (open("/dev/null", O_WRONLY, 0) != 1) {
	    LOGERR_DIE(LOG_ERR) _("Setup_log: open /dev/null failed"));
	}

    /*
     * open logfile; if it is "-", use STDERR; if Foreground is set, use stderr
     */
	if( fstat(2,&statb) == -1 && dup2(1,2) == -1 ){
		LOGERR_DIE(LOG_ERR) _("Setup_log: dup2(%d,%d) failed"), 1, 2);
	}
    if( logfile == 0 ){
		if( !Foreground_LPD && dup2(1,2) == -1 ){
			LOGERR_DIE(LOG_ERR) _("Setup_log: dup2(%d,%d) failed"), 1, 2);
		}
	} else if( safestrcmp(logfile, "-") ){
		close(2);
		if( Checkwrite(logfile, &statb, O_WRONLY|O_APPEND, 0, 0) != 2) {
			LOGERR_DIE(LOG_ERR) _("Setup_log: open %s failed"), logfile );
		}
	}
}

/***************************************************************************
 * Reinit()
 * Reinitialize the database/printcap/permissions information
 * 1. free any allocated memory
 ***************************************************************************/

void Reinit(void)
{
	Reread_config = 1;
	(void) plp_signal (SIGHUP,  (plp_sigfunc_t)Reinit);
}


/***************************************************************************
 * Get_lpd_pid() and Set_lpd_pid()
 * Get and set the LPD pid into the LPD status file
 ***************************************************************************/

int Get_lpd_pid(void)
{
	int pid;
	int lockfd;
	char *path;
	struct stat statb;

	path = safestrdup3( Lockfile_DYN,".", Lpd_port_DYN, __FILE__, __LINE__ );
	pid = -1;
	lockfd = Checkread( path, &statb );
	if( lockfd >= 0 ){
		pid = Read_pid( lockfd, (char *)0, 0  ); 
	}
	if( path ) free(path); path = 0;
	return(pid);
}

void Set_lpd_pid(void)
{
	int lockfd;
	char *path;
	struct stat statb;

	path = safestrdup3( Lockfile_DYN,".", Lpd_port_DYN, __FILE__, __LINE__ );
	lockfd = Checkwrite( path, &statb, O_WRONLY|O_TRUNC, 1, 0 );
	fchmod( lockfd, (statb.st_mode & ~0777) | 0644 );
	if( lockfd < 0 ){
		int euid = geteuid();
		To_euid_root();
		lockfd = Checkwrite( path, &statb, O_WRONLY|O_TRUNC, 1, 0 );
		if( lockfd > 0 ){
			fchown( lockfd, DaemonUID, DaemonGID );
			fchmod( lockfd, (statb.st_mode & ~0777) | 0644 );
		}
		To_euid(euid);
	}
	if( lockfd < 0 ){
		LOGERR_DIE(LOG_ERR) _("lpd: Cannot open '%s'"), path );
	} else {
		int euid = geteuid();
		To_euid_root();
		fchown( lockfd, DaemonUID, DaemonGID );
		fchmod( lockfd, (statb.st_mode & ~0777) | 0644 );
		To_euid(euid);
		/* we write our PID */
		Server_pid = getpid();
		DEBUG1( "lpd: writing lockfile '%s' fd %d with pid '%d'",path,lockfd,Server_pid );
		Write_pid( lockfd, Server_pid, (char *)0 );
	}
	if(path) free(path); path = 0;
	close( lockfd );
}

int Read_server_status( int fd )
{
	int status, count, found, n;
	char buffer[LINEBUFFER];
	char *name;
	fd_set readfds;	/* for select() */
	struct timeval timeval;
	struct line_list l;

	buffer[0] = 0;
	errno = 0;

	DEBUG1( "Read_server_status: starting" );

	Init_line_list(&l);
	while(1){
		FD_ZERO( &readfds );
		FD_SET( fd, &readfds );
		memset(&timeval,0, sizeof(timeval));
		status = select( fd+1,
			FD_SET_FIX((fd_set *))&readfds,
			FD_SET_FIX((fd_set *))0,
			FD_SET_FIX((fd_set *))0, &timeval );
		DEBUG1( "Read_server_status: select status %d", status);
		if( status == 0 ){
			break;
		} else if( status < 0 ){
			close(fd);
			fd = 0;
			break;
		}
		status = read(fd,buffer,sizeof(buffer)-1);
		DEBUG1( "Read_server_status: read status %d", status );
		if( status <= 0 ){
			close(fd);
			fd = 0;
			break;
		}
		buffer[status] = 0;
		/* we split up read line and record information */
		Split(&l,buffer,Whitespace,0,0,0,0,0,0);
		if(DEBUGL1)Dump_line_list("Read_server_status - input", &l );
		for( count = 0; count < l.count; ++count ){ 
			name = l.list[count];
			found = 0;
			for( n = 0;!found && n < Servers_line_list.count; ++n ){
				found = !safestrcasecmp( Servers_line_list.list[n], name);
			}
			if( !found ){
				Add_line_list(&Servers_line_list,name,0,0,0);
			}
		}
	}
	if(DEBUGL2)Dump_line_list("Read_server_status - waiting for start",
			&Servers_line_list );
	return(fd);
}

/***************************************************************************
 * void Get_parms(int argc, char *argv[])
 * 1. Scan the argument list and get the flags
 * 2. Check for duplicate information
 ***************************************************************************/

 static char *msg[] = {
	N_("usage: %s [-FV] [-D dbg] [-L log]\n"),
	N_(" Options\n"),
	N_(" -D dbg      - set debug level and flags\n"),
	N_("                 Example: -D10,remote=5\n"),
	N_("                 set debug level to 10, remote flag = 5\n"),
	N_(" -F          - run in foreground, log to STDERR\n"),
	N_("               Example: -D10,remote=5\n"),
	N_(" -L logfile  - append log information to logfile\n"),
	N_(" -V          - show version info\n"),
	N_(" -p port     - listen on this port\n"),
	0,
};

void usage(void)
{
	int i;
	char *s;
	for( i = 0; (s = msg[i]); ++i ){
		if( i == 0 ){
			FPRINTF( STDERR, _(s), Name);
		} else {
			FPRINTF( STDERR, "%s", _(s) );
		}
	}
	exit(1);
}

 char LPD_optstr[] 	/* LPD options */
 = "D:FL:VX:p:" ;

void Get_parms(int argc, char *argv[] )
{
	int option, verbose = 0;

	while ((option = Getopt (argc, argv, LPD_optstr )) != EOF){
		switch (option) {
		case 'D': Parse_debug(Optarg, 1); break;
		case 'F': Foreground_LPD = 1; break;
		case 'L': Logfile_LPD = Optarg; break;
		case 'V': ++verbose; break;
        case 'X': Worker_LPD = Optarg; break;
		case 'p': Lpd_port_arg = Optarg; break;
		default: usage(); break;
		}
	}
	if( Optind != argc ){
		usage();
	}
	if( verbose ) {
		FPRINTF( STDERR, "%s\n", Version );
		if( verbose > 1 ) Printlist( Copyright, 1 );
		exit(0);
	}
}

int Start_all( int first_scan )
{
	struct line_list args, passfd;
	int pid, p[2];

	Init_line_list(&passfd);
	Init_line_list(&args);

	DEBUG1( "Start_all: begin" );
	Started_server = 0;
	if( pipe(p) == -1 ){
		LOGERR_DIE(LOG_INFO) _("Start_all: pipe failed!") );
	}
	Max_open(p[0]); Max_open(p[1]);
	DEBUG1( "Start_all: fd p(%d,%d)",p[0],p[1]);

	Setup_lpd_call( &passfd, &args );
	Set_str_value(&args,CALL,"all");

	Check_max(&passfd,2);
	Set_decimal_value(&args,INPUT,passfd.count);
	passfd.list[passfd.count++] = Cast_int_to_voidstar(p[1]);
	Set_decimal_value(&args,FIRST_SCAN,first_scan);

	pid = Make_lpd_call( "all", &passfd, &args );

	Free_line_list( &args );
	passfd.count = 0;
	Free_line_list(&passfd);
	close(p[1]);
	if( pid < 0 ){
		close( p[0] );
		p[0] = -1;
	}
	DEBUG1("Start_all: pid %d, fd %d", pid, p[0] );
	return(p[0]);
}

plp_signal_t sigchld_handler (int signo)
{
	signal( SIGCHLD, SIG_DFL );
	write(Lpd_request,"\n", 1);
}

void Setup_waitpid (void)
{
	signal( SIGCHLD, SIG_DFL );
}

void Setup_waitpid_break (void)
{
	(void) plp_signal_break(SIGCHLD, sigchld_handler);
}

void Fork_error( int last_fork_pid_value )
{
	DEBUG1("Fork_error: %d", last_fork_pid_value );
	if( last_fork_pid_value < 0 ){
		LOGMSG(LOG_CRIT)"LPD: fork failed! LPD not accepting any requests");
	}
}
